Add support for the XPT2046 touchscreen controller

numo68 opened the pull request:
## Description:

This pull-request adds a support for the XPT2046 touchscreen controller that is used with many cheap LCD displays. The code is kept simple and just publishes the touch event and the coordinates calibrated and scaled for the display dimensions. The user is free to hook anything needed there. I only tested on my ILI9314/XPT2046 display combo.

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1015

fixes: https://github.com/esphome/feature-requests/issues/1102

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

numo68 committed:
Add support for the XPT2046 touchscreen controller

numo68 commented:
@glmnet @justfalter hmm. not sure what happened here with the probot and light, probably got confused with a borken push that I did trying to rebase, sorry for that. This has nothing to do with light.

glmnet commented on esphome/components/xpt2046/sensor.py:
```diff
@@ -0,0 +1,134 @@
+            cv.Optional(
+                CONF_REPORT_INTERVAL, default="1s"
+            ): cv.positive_time_period_milliseconds,
+            cv.Optional(CONF_TRANSFORM): cv.All(
+                cv.ensure_list(cv.enum(TRANSFORM, upper=True)), cv.Length(max=3)
+            ),
```
can this just be 3 booleans?
Another option will be if invert x / y lets you enter calibration max < min, then let math work

numo68 replied:
Changed this to a boolean and the invert is now handled with max < min, good idea.

glmnet reviewed (commented):
1. Please make this a root component (move from sensor.py to __init__.py)
2. While ESPHome does not have an api to building on touch screens, giving the opportunity to setup buttons as binary sensors will be not too hard, please check mpr121 on how to setup binary sensors so they can have attributes like coordinates and then the sensors can execute actions.
3. What is missing here also is a real callback, instead of using the touch sensor with value 1, that's really not how it should work: you can create a [action trigger ](https://esphome.io/guides/automations.html#config-lambda) passing in coordinate information, then in the c++ code an "advanced" action can be executed, this complements the binary sensors easy of use.

Thanks!

numo68 commented:
Hello @glmnet , many thanks for the review and valuable hints, I am a complete esphome newbie so they are much appreciated.

> 2. While ESPHome does not have an api to building on touch screens, giving the opportunity to setup buttons as binary sensors will be not too hard, please check mpr121 on how to setup binary sensors so they can have attributes like coordinates and then the sensors can execute actions.

Do you mean creating one binary sensor per "virtual" button specified through location coordinates? I considered this, but as I am using quite a few pages with different buttons I was not able to come up with anything that would seem reasonable. I have built some infrastructure with things like momentary, toggle and radio buttons coupling the pages, drawing to a display buffer and areas sensitive to touch, but that is unfortunately not a PR quality by far and probably never will be. I'll think about it but most probably not in this PR.

> 3. What is missing here also is a real callback, instead of using the touch sensor with value 1, that's really not how it should work: you can create a [action trigger ](https://esphome.io/guides/automations.html#config-lambda) passing in coordinate information, then in the c++ code an "advanced" action can be executed, this complements the binary sensors easy of use.

Ah, that seems indeed a better way. Is an `on_state` with `x` being a structure with coordinates and the touched / released boolean the right idiom to use or is it better to create a new one like `on_touch` / `on_release` ?

Thanks

glmnet commented:
2. Yes. Still apply because The button lambda could check display active page to decide if the action would make sense. 

3.  You can pass x and y as two variables. No need for structure

numo68 committed:
xpt2046: change from sensor to action

numo68 committed:
xpt2046: rework transform

The swap_x_y is now a boolean
The invert_x/y is deduced from min > max

numo68 committed:
xpt2046: style fixes

numo68 commented on esphome/components/xpt2046/xpt2046.cpp:
```diff
@@ -0,0 +1,180 @@
+
+void XPT2046Component::setup() {
+  if (this->tirq_pin_ != nullptr) {
+    this->tirq_pin_->setup();
```
I checked the datasheet and experiences other developers made. The xpt2046 handling of the pin is weird, pulling it low and high when an access to some registers is performed, meaning that the interrupt handling is far from straightforward and is probably meant more to pull the MCU from a deep sleep than to be used to while in the normal operation.

Maybe I'll try to use the value of the pin if configured, at least saving on unnecessary SPI transactions when noone is touching the thing. For now I have removed it.

numo68 committed:
xpt2046: add virtual buttons

numo68 committed:
xpt2046: style fixes

numo68 committed:
xpt2046: implement the optional penirq pin handling

numo68 commented:
Hello @glmnet . I updated the code
- the code style issues are (hopefully( resolved
- there is on_state with x, y and touched arguments
- there are virtual buttons defining the sensitive area via attributes
- the swap_x_y is now a boolean and the inversions are configurable through max < min
- the penirq pin (that's the name from the datasheet) is optional and if present, speeds up calling the update() right from the loop() and saves unnecessary SPI transactions when not touched. Due to the weirdness of the chip that is messing with the pin while doing the conversions I did not attempt to do a real interrupt handling

I did not update the documentation PR in the docs repo yet.

Would you please review this again? Many thanks in advance.

numo68 committed:
xpt2046: change threshold validation

glmnet commented:
I would rename `penirq` to `irq` or in any case to `pen_irq` my board says "T_IRQ" this can be value information for docs (also pins are T_DO for MISO, DIN for MOSI, T_CLK and T_CS. I'm using same SPI bus for touch and ili9341 for testing

This looks a lot better for me, the binary sensor worked right away, the `on_state` too, but I had to create global variables. 
```yaml

globals:
  - id: touch_x
    type: int
  - id: touch_y
    type: int

  - id: touch_on
    type: boolean

xpt2046:
  cs_pin: GPIO12
  on_state:
    - lambda: |-
        id (touch_on) = touched;
        id(touch_x) = x;
        id(touch_y) = y;
    - component.update: display_x

display:
  - platform: ili9341
    lambda: |-
      it.fill(BLACK);
      if (id(esphome_logo_touch).state)
        it.image(0, 0, id(esphome_logo));
      else
        it.image(0, 0, id(esphome_logo), RED, BLACK);
      if (id(touch_on))
        it.filled_circle(id(touch_x) - 5, id(touch_y) - 5, 10, RED);

```

I believe we can have a simpler interface here, maybe adding public members on the class itself so they can be easily accessed from the display lambda, e.g.

```yaml
# no need to declare globals

xpt2046:
  id: touch
  cs_pin: GPIO12
  on_state:
    # here we just trigger a display update
    - component.update: display_x

display:
  - platform: ili9341
    lambda: |-
      it.fill(BLACK);
      if (id(esphome_logo_touch).state)
        it.image(0, 0, id(esphome_logo));
      else
        it.image(0, 0, id(esphome_logo), RED, BLACK);
      if (id(touch).touched)
        it.filled_circle(id(touch).x - 5, id(touch).y - 5, 10, RED);

```

Did test max < min and worked ok, didn't test swap (a 90º rotation i.e.) but I believe this method is ok
Didn't test the irq yet, will do. Reaction time is quite good with 50ms poll time though, I believe the render itself takes 50ms more

Also I'd like to know your use and test cases, I'm doing just tests so can't tell best usage methods.
Those should be given in examples too

numo68 commented:
Hi, thanks for the review. I'll address the comments in the evening or next days and I'll also try to write some documentation with examples, way to calibrate etc.

Regarding the render, maybe you can try to take a look at #1628 . This is what I did exactly because the render time interfered with the responsiveness.

I don't have any dedicated test cases (yet). My "big picture" use case is a gui to my home assistant installation with the only hardware added being a pressure sensor. The goal is to see everything and control some important things, i.e. a few pages with the control elements from page change buttons to some momentary (blinds up), toggles (1 to n room selection) and radio-ones (auto / manual / vacation mode for thermostats) and later some numeric input keyboard (start/end time for the vacation mode). Generally a dimmed display, touch first, the main page brightens, touch again and you get to a set of navigable pages. See examples at the end.

What I am doing now is directly feeding the touch events to a class handling the gui. This class is only implemented as a custom component at the project level, not in the repository; I'll probably change that as the advantages of the yaml configuration and python code generation are really convenient and they don't seem to be available for the custom components. Unfortunately not a PR quality by far.

I'll add the public class variables but I'm inclined to leaving the `x`, `y`, `touched` arguments implemented as well. Maybe I am overthinking it but one case where they are needed would be e.g. 
``` on_state:
    - lambda: '... use x, y'
    - delay 1s
    - lambda '... use x, y'
 ```
I admit that I just prefer function arguments to variables where possible (my main job has a lot to do with threads and asynchronous things so probably that's why :) ). I guess that the added code space is negligible. What do you think? I'll add the raw values as well, that would make getting the calibration values possible without needing to mess with the debug level and looking at fast scrolling logs.

![IMG_0894 (Custom)](https://user-images.githubusercontent.com/5095188/112447153-46863500-8d51-11eb-9881-8d5fe24ac97c.jpg) ![IMG_0895 (Custom)](https://user-images.githubusercontent.com/5095188/112447171-4ab25280-8d51-11eb-97c3-5a91dea4844a.jpg)

numo68 committed:
xpt2046: change penirq to irq, add public members

numo68 commented:
I renamed penirq to irq (my board uses the T_IRQ description as well) and added `x`, `y`, `touched`, `x_raw`, `y_raw` and `z_raw` public members. I left the action arguments `x`, `y` and `touched` as well.

My usage is now something in the line of 
```yaml
xpt2046:
  id: "touchscreen"
  cs_pin: 17
  irq_pin: 16
  update_interval: 50ms
  dimension_x: 240
  dimension_y: 320
  calibration_x_min: 3860
  calibration_x_max: 280
  calibration_y_min: 340
  calibration_y_max: 3860
  swap_x_y: False
  threshold: 400
  #report_interval: 1s

  on_state:
    - lambda: |-
        ESP_LOGI("main", "xpt2046 x=%d, y=%d, touched=%d, x_raw=%d, y_raw=%d, z_raw=%d",
          id(touchscreen).x,
          id(touchscreen).y,
          (int) id(touchscreen).touched,
          id(touchscreen).x_raw,
          id(touchscreen).y_raw,
          id(touchscreen).z_raw
          );
    - if:
        condition:
          lambda: 'return touched;'
        then:
          - if:
              condition:
                lambda: 'return id(active_time_remaining) == 0;'
              then:
                - light.turn_on:
                    id: 'back_light'
                    transition_length: 0s
                    brightness: !lambda 'return id(active_brightness);'
              else:
                - lambda: 'gui->touch_press(x, y);'
          - globals.set:
              id: active_time_remaining
              value: !lambda 'return id(idle_timeout);'
        else:
          - lambda: 'gui->touch_release();'

interval:
  - interval: 5s
    then:
      - if:
          condition:
            lambda: 'return id(active_time_remaining) > 5;'
          then:
            - lambda: 'id(active_time_remaining) -= 5;'
          else:
            - if:
                condition:
                  lambda: 'return id(active_time_remaining) > 0;'
                then:
                  - display.page.show: page_main
                  - component.update: tft
                  - light.turn_on:
                      id: 'back_light'
                      transition_length: 2s
                      brightness: !lambda 'return id(idle_brightness);'
                  - globals.set:
                      id: active_time_remaining
                      value: '0'

globals:
  # Time to remain active after the last interaction
  - id: idle_timeout
    type: int
    restore_value: no
    initial_value: '30'
  # Remaining time to go back to idle in seconds
  - id: active_time_remaining
    type: int
    restore_value: no
    initial_value: '0'
  # Brightness when idle
  - id: idle_brightness
    type: float
    restore_value: no
    initial_value: '0.33'
  # Brightness when activated
  - id: active_brightness
    type: float
    restore_value: no
    initial_value: '1'

```
where the gui is my button-processing class, tft is an ili9341 display and light a PWM controlled backlight.

jesserockz commented on esphome/components/xpt2046/__init__.py:
```diff
@@ -0,0 +1,124 @@
+from esphome.const import CONF_ID, CONF_ON_STATE, CONF_THRESHOLD, CONF_TRIGGER_ID
+
+CODEOWNERS = ["@numo68"]
+AUTO_LOAD = ["spi", "binary_sensor"]
```
This should be a dependency, not an auto-load because SPI requires configuration and cannot be auto loaded.
```suggestion
AUTO_LOAD = ["binary_sensor"]
DEPENDENCIES = ["spi"]
```

glmnet commented:
> I'll add the public class variables but I'm inclined to leaving the `x`, `y`, `touched` arguments implemented as well.

Absolutely, I did not mean to ask for removal of those. 


It'd be nice if you try to get your UI working with yaml, you can may be simplify the timer thing with a [timer script](https://esphome.io/guides/automations.html#timers-and-timeouts) also a 

```yaml
  condition:
    display.active_page: 2
```

can be added (in another pr)

numo68 committed:
Update esphome/components/xpt2046/__init__.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

numo68 commented:
I have added the documentation in the esphome/esphome-docs#1015

numo68 committed:
xpt2046: allow 'never' for report_interval

glmnet commented on esphome/components/xpt2046/xpt2046.h:
```diff
@@ -0,0 +1,83 @@
+  void loop() override;
+  void update() override;
+
+  int16_t x{0}, y{0};
```
Documenting these will be appreciated. The api docs are generated and accessible in the page.

numo68 replied:
done

glmnet reviewed (approved):
Looks good to me

numo68 committed:
xpt2046: document the public interface, fix comment

numo68 committed:
xpt2046: style fixes

numo68 committed:
xpt2046: log the irq pin in dump_config

lyazide commented:
I tried this on my hassio. All is working like a charm. Thank you for this. I will try to post my setup next week for reference.

numo68 committed:
add xpt2046 test

numo68 committed:
move the test to test4 because of space

PiratWeber commented:
Hey! Thanks for creating this feature! I can't wait till it's merged and released!
